### ISSUE TITLE
A "urn:oid:1.3.6.1.4.1.5923.1.1.1.10" (EPTI) attribute value must be a NameID, none found for value no. "0"

### ISSUE THREAD

**Issue by impressio** [simplesamlphp/saml2#105]
Running SimpleSAMLphp 1.15 (RC1/2) as SP, against older version SimpleSAMLphp 1.8 as IdP. Gets the following "Received message" which fails at "SAML2\Assertion::parseAttributeValue".

> 12:03:57 DEBUG     <<saml:Subject>>
> 12:03:57 DEBUG       <saml:NameID SPNameQualifier="sp/" Format="urn:oasis:names:tc:SAML:2.0:nameid-format:persistent">---identical---</saml:NameID>
> 12:03:57 DEBUG       <saml:SubjectConfirmation Method="urn:oasis:names:tc:SAML:2.0:cm:bearer">
> 12:03:57 DEBUG         <saml:SubjectConfirmationData NotOnOrAfter="2017-10-27T10:08:56Z" Recipient="sp/module.php/saml/sp/saml2-acs.php/default-sp" InResponseTo="_snip"/>
> 12:03:57 DEBUG       </saml:SubjectConfirmation>
> 12:03:57 DEBUG     </saml:Subject>
> 
> 12:03:57 DEBUG     <<saml:AttributeStatement>>
> 12:03:57 DEBUG       <saml:Attribute Name="urn:oid:1.3.6.1.4.1.5923.1.1.1.6" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri">
> 12:03:57 DEBUG         <saml:AttributeValue xsi:type="xs:string”>snip@snip.snip</saml:AttributeValue>
> 12:03:57 DEBUG       </saml:Attribute>
> 12:03:57 DEBUG       <saml:Attribute Name="urn:oid:2.5.4.3" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri">
> 12:03:57 DEBUG         <saml:AttributeValue xsi:type="xs:string”>g s</saml:AttributeValue>
> 12:03:57 DEBUG       </saml:Attribute>
> 12:03:57 DEBUG       <saml:Attribute Name="urn:oid:1.3.6.1.4.1.5923.1.1.1.10" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri">
> 12:03:57 DEBUG         <saml:AttributeValue xsi:type="xs:string">---identical---</saml:AttributeValue>
> 12:03:57 DEBUG       </saml:Attribute>
> 12:03:58 DEBUG     </saml:AttributeStatement>
> 12:03:58 DEBUG   </saml:Assertion>
> 12:03:58 DEBUG </samlp:Response>
> 12:03:58 ERROR SimpleSAML_Error_Error: UNHANDLEDEXCEPTION
> 12:03:58 ERROR Backtrace:
> 12:03:58 ERROR 1 simplesamlphp\www\_include.php:45 (SimpleSAML_exception_handler)
> 12:03:58 ERROR 0 [builtin] (N/A)
> 12:03:58 ERROR Caused by: SAML2\Exception\RuntimeException: A "urn:oid:1.3.6.1.4.1.5923.1.1.1.10" (EPTI) attribute value must be a NameID, none found for value no. "0"
> 12:03:58 ERROR Backtrace:
> 12:03:58 ERROR 7 simplesamlphp\vendor\simplesamlphp\saml2\src\SAML2\Assertion.php:539 (SAML2\Assertion::parseAttributeValue)
> 12:03:58 ERROR 6 simplesamlphp\vendor\simplesamlphp\saml2\src\SAML2\Assertion.php:521 (SAML2\Assertion::parseAttributes)
> 12:03:58 ERROR 5 simplesamlphp\vendor\simplesamlphp\saml2\src\SAML2\Assertion.php:280 (SAML2\Assertion::__construct)
> 12:03:58 ERROR 4 simplesamlphp\vendor\simplesamlphp\saml2\src\SAML2\Response.php:38 (SAML2\Response::__construct)
> 12:03:58 ERROR 3 simplesamlphp\vendor\simplesamlphp\saml2\src\SAML2\Message.php:578 (SAML2\Message::fromXML)
> 12:03:58 ERROR 2 simplesamlphp\vendor\simplesamlphp\saml2\src\SAML2\HTTPPost.php:76 (SAML2\HTTPPost::receive)
> 12:03:58 ERROR 1 simplesamlphp\modules\saml\www\sp\saml2-acs.php:31 (require)
> 12:03:58 ERROR 0 simplesamlphp\www\module.php:135 (N/A)

`        $nameId = Utils::xpQuery(
            $subject,
            './saml_assertion:NameID | ./saml_assertion:EncryptedID/xenc:EncryptedData'
        );`

[https://github.com/simplesamlphp/saml2/blob/708e6785ed2ea046e971bb86f62a9d33d4438aa5/src/SAML2/Assertion.php#L291](url)

`$eptiNameId = Utils::xpQuery($eptiAttributeValue, './saml_assertion:NameID');`

[https://github.com/simplesamlphp/saml2/blob/708e6785ed2ea046e971bb86f62a9d33d4438aa5/src/SAML2/Assertion.php#L529](url)

**Comment by thijskh** [comment]
You are sending a string in the eduPersonTargetedID (urn:oid:1.3.6.1.4.1.5923.1.1.1.10) attribute. That is not allowed; per the specification eduPersonTargetedID is an XML construct, being a NameID. Your IdP needs to change what it sends for the eduPersonTargetedID.

**Comment by impressio** [comment]
@thijskh: Is there anything we can do on our SP side? And for the IdP, how do they change this setting (and specifically for our SP)? Thanks.

**Comment by impressio** [comment]
We can login when we tested disable NameID lookup in "private function parseAttributeValue($attribute, $attributeName)" (tested with this particular IdP). Is it purely a string vs XML issue involved?

**Comment by impressio** [comment]
Should I expect the IdP to send the attribute in this way instead?

```xml
<saml:Attribute Name="urn:oid:1.3.6.1.4.1.5923.1.1.1.10"
                NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri">
    <saml:AttributeValue>
        <saml:NameID NameQualifier="IdP-Entity-ID"
                     SPNameQualifier="SP-Entity-ID">
            xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
        </saml:NameID>
    </saml:AttributeValue>
</saml:Attribute>
```

**Comment by thijskh** [comment]
Yes, that is how it should look.
Not sure what your IdP config looks like, but if you're running 1.8 that might be a good idea to upgrade anyway.

**Comment by impressio** [comment]
I have no control over the IdP wich belongs to an institution. It is one of the IdPs we get from a federation metadata stream. We spent 1.5 hours on the phone thisfriday afternoon, trying to solve this. If the guy has a boss (which I think he has), I would not expect any futher help. However, we would like to know how how to prevent it happening in the future.

Is below what the IdP need to send?

```php
'attributeencodings' => array('urn:oid:1.3.6.1.4.1.5923.1.1.1.10' => 'raw',),
```

Is below a better approach? Will this move the NameID from Subject to AttributeValue as RAW XML? 

```php
'authproc' => array(
    // Generate the persistent NameID.
    2 => array(
        'class' => 'saml:PersistentNameID',
        'attribute' => 'eduPersonPrincipalName',
    ),
    // Add the persistent to the eduPersonTargetedID attribute
    60 => array(
        'class' => 'saml:PersistentNameID2TargetedID',
        'attribute' => 'eduPersonTargetedID', // The default
        'nameId' => TRUE, // The default
    ),
    // Use OID attribute names.
    90 => array(
        'class' => 'core:AttributeMap',
        'name2oid',
    ),
),
// The URN attribute NameFormat for OID attributes.
'attributes.NameFormat' => 'urn:oasis:names:tc:SAML:2.0:attrname-format:uri',
'attributeencodings' => array(
    'urn:oid:1.3.6.1.4.1.5923.1.1.1.10' => 'raw', /* eduPersonTargetedID with oid NameFormat is a raw XML value */
),
```

Is it possible for the IdP choose to send the RAW XML specifically to our SP? Is it possible for our SP to convert the "received message" to RAW XML. Also I can't see 'NameQualifier' from this specific IdP. And at last, is version 1.8 capable and compatible with the current version and protocols?

@thijskh: Many thanks for you guidance.

**Comment by thijskh** [comment]
I believe the approach with PersistentNameID2TargetedID sketched above should work. However, that authproc filter has only been added in SSP 1.11. Not sure how they generate the attribute now. Core of the problem is indeed that according to specification the attribute value _must_ be an XML nameID.

It might be difficult to fix this in the SP because the error already happens at parsing the assertion.

Some possible alternative approaches:
- IdP stops releasing eduPersonTargetedID at all (to your SP or in general). You can get the same information from the Subject NameID.
- You could try and see if you disable the code in saml2 that special cases eduPersonTargetedID makes it work again. (https://github.com/simplesamlphp/saml2/blob/v3.0.2/src/SAML2/Assertion.php#L534-L550) However. that would of course create compatibility problems with other IdPs that send the 'correct' value to you. Or you could change the code to handle both the legacy and the correct case.

**Comment by impressio** [comment]
@thijskh: **Many thanks** for your help. Will follow your advice.

However, I don't know if I am still mixing up our problem with this one: [**eduPersonTargetedId with SSP 1.14.6**](https://groups.google.com/forum/#!topic/simplesamlphp/Zn9KFqflUjU)

**Comment by thijskh** [comment]
That was fixed in 1.14.7 (2016) already.

**Comment by impressio** [comment]
Confirming "IdP stops releasing eduPersonTargetedID at all (to your SP or in general)" is a working solution.

Many thanks @thijskh
